src/workload.rs

```diff
@@ -512,59 +620,173 @@ impl WorkloadInformation {
         }
     }
 
-    pub async fn find_upstream(&self, addr: SocketAddr, hbone_port: u16) -> Option<Upstream> {
-        self.fetch_address(&addr).await;
+    pub async fn find_upstream(
+        &self,
+        network: &str,
+        addr: SocketAddr,
+        hbone_port: u16,
+    ) -> Option<Upstream> {
+        self.fetch_address(network_addr(network, addr.ip())).await;
         let wi = self.info.lock().unwrap();
-        wi.find_upstream(addr, hbone_port)
+        wi.find_upstream(network, addr, hbone_port)
+    }
+
+    pub async fn find_waypoint(&self, wl: Workload) -> Result<Option<Upstream>, WaypointError> {
+        if let Some(gw_address) = &wl.waypoint {
+            // Even in this case, we are picking a single upstream pod and deciding if it has a remote proxy.
+            // Typically this is all or nothing, but if not we should probably send to remote proxy if *any* upstream has one.
+            let wp_nw_addr = match &gw_address.destination {
+                gatewayaddress::Destination::Address(ip) => ip,
+                gatewayaddress::Destination::Hostname(_) => {
+                    info!(%wl.name, "waypoint hostname lookup not supported yet");
+                    return Ok(None); // TODO: implement unsupported feature
```

Comment:
```suggestion
                    return Err(WaypointError::FindWaypointError);
```
?

Reply:
addressed in https://github.com/istio/ztunnel/pull/486/commits/ecc06af22e028572215f01f205db579f13801728